### DISCUSSION TITLE
How to determine if commit exists in one or more remotes?

### DISCUSSION THREAD

**Discussion by lyndsey-ferguson** [go-git/go-git#1149] [Q&A] [answered]
I have a task to check if a given commit has been pushed up to 1 or more remotes. Using the `git` CLI with Python, we would run `        github_sha_cmd = shlex.split(f'git ls-remote git@github.com:{fork}/{repository} -h refs/heads/{branch}')` and then compare it to the given `sha`.

I also see that when I run `git log`, the output shows:
```
commit 96999f37a590b4b283a04f8346ae69bda813ef33 (HEAD -> mybranch, lyndsey-ferguson/mybranch, otherremote/mybranch)
```

So, git seems to actually know if this commit is on those branches, I would love to be able to leverage that so as to not make more calls.

How can I do something similar with the go-git SDK?

**Comment by pjbgf** [comment, marked as answer]
`Remote.List` is what you are looking for. The existing [tests](https://github.com/go-git/go-git/blob/4d0c2cccb890be585b74fea3805db6e555f42cfc/remote_test.go#L666-L668) might give you an idea on how to use it.

**Reply by lyndsey-ferguson** [reply]
Would the work flow like this? 

```
listOptions := ListOptions{}

refs, err := remote.List(&listOptions)

var branchRef *plumbing.Reference
var gitHubSha string

for _, rf := range refs {
  if rf.Name().String() == "refs/head/mybranch" {
    branchRef = rf
    break
  }
}

if branchRef != nil && branchRef.Hash().String() == desiredSha {
  return True
}
return False
```

**Reply by lyndsey-ferguson** [reply]
Here's what I created:

```go
var (
  ErrShaNotFoundOnGitHubRemote    = errors.New("SHA not found on GitHub remote")
)

var getRemoteReferences = func(remote *git.Remote) ([]*plumbing.Reference, error) {
	listOptions := git.ListOptions{}
	refs, err := remote.List(&listOptions)
	if err != nil {
		return nil, err
	}
	return refs, nil
}

func validateShaAtReferenceOnRemote(localSha string, localRef string, remote *git.Remote) error {
	refs, err := getRemoteReferences(remote)
	if err != nil {
		return nil
	}
	var branchRef *plumbing.Reference

	for _, rf := range refs {
		if rf.Name().String() == localRef {
			branchRef = rf
			break
		}
	}

	if branchRef != nil && branchRef.Hash().String() == localSha {
		return nil
	}
	return ErrShaNotFoundOnGitHubRemote
}
```

**NOTE**: I made getRemoteReferences function a variable that I can switch out for mocks during my tests
